vault/expiration.go

```diff
@@ -1024,19 +1044,20 @@ func (m *ExpirationManager) Renew(ctx context.Context, leaseID string, increment
 		}
 	}
 
-	{
-		m.pendingLock.Lock()
-		if err := m.persistEntry(ctx, le); err != nil {
-			m.pendingLock.Unlock()
-			return nil, err
-		}
-
-		// Update the expiration time
-		m.updatePendingInternal(le)
-		m.pendingLock.Unlock()
+	// TODO: there is a race condition here with revocation, which may
+	// begin after we have stopped the entry timer, but finish before we get here
+	// and persist the entry.
+	//
+	// The stopEntry() call above protects against concurrent renewals; only one
+	// will be able to write the entry. Extending the lock to cover persistEntry,
+	// as was previously the case, would create more contention, but not solve
+	// the race with revocation.
+	if err := m.persistEntry(ctx, le); err != nil {
```

Comment:
Do we want to clone `le` so we can decide to put the old or new value back into pending depending on if we return an error?

Reply:
If the renewEntry call succeeds, but persistEntry fails, which version do we want to have included in-memory?

My guess is that after renewEntry succeeds then some persistent change might already have been made.  I see in the documentation that the secret is supposed to be the "same", but I'm unclear what a secret backend is actually allowed/supposed to do with a RenewOperation.

Reply:
I did add the clone, but let me know what your opinion is on failures between renewEntry and persistEntry.